pkg/policy/consumer.go

```diff
@@ -205,44 +179,32 @@ func (c *Consumable) AllowIngressIdentityLocked(cache *ConsumableCache, id Numer
 			"consumable":       logfields.Repr(c),
 		}).Debug("Allowing security identity on ingress for consumable")
 		c.addToMaps(id)
-		c.IngressIdentities[id] = true
-		return true
-	}
 
-	c.IngressIdentities[id] = true
-
-	return false // not changed.
-}
-
-// AllowIngressIdentityAndReverseLocked adds the given security identity to the
-// Consumable's IngressIdentities map and BPF policy map, as well as this
-// Consumable's security identity to the Consumable representing id's Ingress
-// Identities map and its BPF policy map.
-// Must be called with Consumable mutex Locked.
-// Returns true if changed, false if not.
-func (c *Consumable) AllowIngressIdentityAndReverseLocked(cache *ConsumableCache, id NumericIdentity) bool {
-	log.WithFields(logrus.Fields{
-		logfields.Identity + ".from": id,
-		logfields.Identity + ".to":   c.ID,
-	}).Debug("Allowing direction")
-	changed := c.AllowIngressIdentityLocked(cache, id)
+		// If id corresponds to a reserved identity, Consumable corresponding to
+		// that security identity needs to be updated explicitly, as reserved
+		// identities do not have a corresponding endpoint for which policy
+		// recalculation (when Consumables are updated) is done.
+		if id.IsReservedIdentity() {
+			reservedConsumable := cache.Lookup(id)
+			if reservedConsumable != nil {
+				// Avoid deadlock ; is this necessary? Being cautious.
+				if id != c.ID {
+					reservedConsumable.Mutex.Lock()
+					reservedConsumable.AllowIngressIdentityLocked(cache, c.ID)
+					reservedConsumable.Mutex.Unlock()
+				} else {
+					reservedConsumable.AllowIngressIdentityLocked(cache, c.ID)
```

Comment:
Why is the logic exactly the same on both branches?

Reply:
It is not the same; one acquires and releases the lock, and one does not. If a Consumable can allow itself (which it can IIRC), then we should not call the lock for it because it will already be locked by the caller of `AllowIngressIdentityLocked`.

Reply:
Please add this comment in the code

Reply:
Done.

Reply:
Sorry, now it is done - didn't push to origin for the branch before I commented.